datawig/hpo.py

```diff
@@ -0,0 +1,313 @@
+                             label_encoders=label_column,
+                             output_path=self.output_path)
+
+        hp_imputer.fit(train_df=train_df,
+                       test_df=test_df,
+                       ctx=get_context(),
+                       learning_rate=global_parms['learning_rate'],
+                       num_epochs=global_parms['num_epochs'],
+                       patience=global_parms['patience'],
+                       test_split=.1,
+                       weight_decay=global_parms['weight_decay'],
+                       batch_size=global_parms['batch_size'],
+                       final_fc_hidden_units=global_parms['final_fc_hidden_units'],
+                       calibrate=True)
+
+        # add suitable metrics to hp series
+        imputed = hp_imputer.predict(test_df)
+        true = imputed[self.imputer.output_column]
+        predicted = imputed[self.imputer.output_column + '_imputed']
+        if is_numeric_dtype(train_df[self.imputer.output_column]):
+            hp['mse'] = mean_squared_error(true, predicted)
+            confidence = np.nan
+        else:
+            confidence = imputed[self.imputer.output_column + '_imputed_proba']
+            hp['f1_micro'] = f1_score(true, predicted, average='micro')
+            hp['f1_macro'] = f1_score(true, predicted, average='macro')
+            hp['precision_micro'] = f1_score(true, predicted, average='micro')
+            hp['precision_macro'] = f1_score(true, predicted, average='macro')
+            hp['recall_micro'] = recall_score(true, predicted, average='micro')
+            hp['recall_macro'] = recall_score(true, predicted, average='macro')
+            hp['coverage_at_80'] = (confidence > .8).mean()
+            hp['coverage_at_90'] = (confidence > .9).mean()
+            # TODO: add support for user provided functions
+        for uds in user_defined_scores:
+            hp[uds[1]] = uds[0](true=true, predicted=predicted, confidence=confidence)
+
+        return hp
+
+    def tune(self,
+             train_df: pd.DataFrame,
+             test_df: pd.DataFrame = None,
+             strategy: str = 'random',
+             num_evals: int = 10,
+             user_defined_scores: list = None):
+
+        """
+        Do grid search or random search for hyperparameter configurations. This method can not tune tfidf vs hashing
+        vectorization but uses tfidf. Also parameters of the output column encoder are not tuned.
+        
+        :param train_df: training data as dataframe
+        :param test_df: test data as dataframe; if not provided, a ratio of test_split of the
+                          training data are used as test data
+
+        :param strategy: 'random' for random search or 'grid' for exhaustive search
+        :param num_evals: number of evaluations for random search
+        :param user_defined_scores: list with entries (Callable, str), where callable is a function
```

Comment:
maybe add one example, or list the default scoring?

Reply:
is it possible to change the user defined function signatures to take parameters named, as in `udf(true, predicted, confidence)`?